pkg/tsdb/cloudwatch/request_parser.go

```diff
@@ -51,32 +53,28 @@ func (e *cloudWatchExecutor) parseQueries(queries []backend.DataQuery, startTime
 	return requestQueries, nil
 }
 
-// migrateLegacyQuery migrates queries that has a `statistics` field to use the `statistic` field instead.
-// This migration is also done in the frontend, so this should only ever be needed for alerting queries
-// In case the query used more than one stat, the first stat in the slice will be used in the statistic field
-// Read more here https://github.com/grafana/grafana/issues/30629
-func migrateLegacyQuery(queries []backend.DataQuery, startTime time.Time, endTime time.Time) ([]*backend.DataQuery, error) {
+// migrateLegacyQuery is also done in the frontend, so this should only ever be needed for alerting queries
+func migrateLegacyQuery(queries []backend.DataQuery, dynamicLabelsEnabled bool, startTime time.Time, endTime time.Time) ([]*backend.DataQuery, error) {
 	migratedQueries := []*backend.DataQuery{}
 	for _, q := range queries {
 		query := q
-		model, err := simplejson.NewJson(query.JSON)
+		queryJson, err := simplejson.NewJson(query.JSON)
 		if err != nil {
 			return nil, err
 		}
 
-		_, err = model.Get("statistic").String()
-		// If there's not a statistic property in the json, we know it's the legacy format and then it has to be migrated
+		if err := migrateStatisticsToStatistic(queryJson); err != nil {
+			return nil, err
+		}
+
+		_, labelExists := queryJson.CheckGet("label")
+		if !labelExists && dynamicLabelsEnabled && queryJson != nil {
```

Comment:
nit: you don't need `queryJson != nil` since you are already doing `queryJson.CheckGet` above (which would panic btw if queryJson is nil)

Reply:
good point, done!